src/rgw/driver/rados/rgw_notify.cc

```diff
@@ -1201,23 +1215,43 @@ reservation_t::reservation_t(const DoutPrefixProvider* _dpp,
 }
 
 reservation_t::reservation_t(const DoutPrefixProvider* _dpp,
-			     rgw::sal::RadosStore* _store,
-			     rgw::sal::Object* _object,
-			     rgw::sal::Object* _src_object,
-			     rgw::sal::Bucket* _bucket,
-			     const std::string& _user_id,
-			     const std::string& _user_tenant,
-			     const std::string& _req_id,
-			     optional_yield y) :
-    dpp(_dpp), store(_store), s(nullptr), size(0) /* XXX */,
-    object(_object), src_object(_src_object), bucket(_bucket),
-    object_name(nullptr),
-    metadata_fetched_from_attributes(false),
-    user_id(_user_id),
-    user_tenant(_user_tenant),
-    req_id(_req_id),
-    yield(y)
-{}
+                             rgw::sal::RadosStore* _store,
+                             rgw::sal::Object* _object,
+                             rgw::sal::Object* _src_object,
+                             rgw::sal::Bucket* _bucket,
+                             const std::string& _user_id,
+                             const std::string& _user_tenant,
+                             const std::string& _req_id,
+                             optional_yield y)
+    : dpp(_dpp),
+      store(_store),
+      s(nullptr),
+      size(0) /* XXX */,
+      object(_object),
+      src_object(_src_object),
+      bucket(_bucket),
+      object_name(nullptr),
+      metadata_fetched_from_attributes(false),
+      user_id(_user_id),
+      user_tenant(_user_tenant),
+      req_id(_req_id),
+      yield(y) {
+  if (store->svc()->zone->get_zonegroup().supports(
+          rgw::zone_features::notification_v2)) {
+    //  for non-request caller (e.g., lifecycle, ObjectSync), bucket attrs
+    //  are not loaded, so force to reload the bucket, that reloads the attr.
+    std::unique_ptr<rgw::sal::Bucket> reload_bucket;
+    int ret = store->load_bucket(
+        dpp, rgw_bucket(bucket->get_tenant(), bucket->get_name()),
+        &reload_bucket, null_yield);
+    if (ret < 0) {
+      ldpp_dout(dpp, 1) << "failed to get_bucket for " << bucket->get_name()
+                        << dendl;
+      return;
+    }
+    bucket = reload_bucket.release();
```

Comment:
so throwing error/exception is a NO_OP coz for this constructor is called for Non-S3 op's (lifecycle, object synced) and if any error comes out of notification for these events, they are not reported up but swallowed. So after loading bucket if there is any error we could ignore

Reply:
maybe we can call `load_bucket()` inside `publish_reserve()` if `attrs` is empty?

Reply:
publish_reserve() is called for both s3 and non-s3 op's and reloading the bucket there would mean we are reloading for all scenario's, 
is there a way to distinguish the call between s3 and non-s3 inside publish_reserve ? if not then i feel this is the right place to reload the bucket

Reply:
we can reload the bucket based on whether `attrs` are empty.
there could eb a case where it is an s3-op and there are no bucket attributes, but it would be most likely that this is becuase the attributes were never loaded

Reply:
> we can reload the bucket based on whether `attrs` are empty. there could eb a case where it is an s3-op and there are no bucket attributes, but it would be most likely that this is becuase the attributes were never loaded

for S3 operations, notification currently supports ObjectCreate and ObjectDelete and both of then need a bucket to be loaded as they are bucket specific operatioin, so there will never be a case that bucket attrs are not loaded for a S3-op.

Coming to loading the attrs, if they are empty, there could be certain buckets that do not have attrs at all, and publish reserve is in the HOT path, so we end up loading bucket atts all the time if we put in publish_reserve.

I feel if bucket->reload fails in notification its not that big a problem (it might not even reach there) , coz there will be more broader issues if bucket->reload() as for all bucket operations we almost load the bucket.

Reply:
there is actually a much better indication: `res.s == nullptr`  9a pointer to `req_state`)

Reply:
> there is actually a much better indication: `res.s == nullptr` (a pointer to `req_state`)

done.